crates/rpc/src/lib.rs

```diff
@@ -73,7 +79,11 @@ impl RpcServer {
             .set_middleware(tower::ServiceBuilder::new()
                 .option_layer(self.cors)
                 .map_result(middleware::versioning::try_map_errors_to_responses)
-                .filter_async(|result| async move {
+                .filter_async(|result: Request<Body>| async move {
+					// skip method_name checks for websocket handshake
+					if result.headers().get("sec-websocket-key").is_some() {
+						return Ok(result);
+					}
```

Comment:
Am I understanding correctly that this means all paths / routes will react to `starknet_subscribe_newHeads`? We will likely want a lot more control than that, as it is likely that the responses will change i.e. we need versioning, or at least some way of controlling where this can be submitted.

Reply:
I might be missing the mark here. But It seemed to check of subscription method from body JSON. But during the WS handshake body is empty, which cause a MALFORMED 400 error. So just added this to skip the `method_name` checks for WS handshake.

Reply:
TL;DR
Pathfinder versions RPC API based on the route that you specify, with the default route pointing to `v0.2`.

So ideally you'd want to be able to subscribe only on the `/rpc/v0.1/pathfinder` route, which is meant for new/experimental features/methods for example:
```
ws://localhost:9545/rpc/pathfinder/v0.1
```
while other routes, including the default (eg: `ws://localhost:9545/`) would fail.

-----------------

A word about how we currently do RPC versioning:

For HTTP transport the RPC api versioning is based on routes (`/, /rpc/v0.2, /rpc/pathfinder/v0.1, /rpc/v0.3`) and depending on the route the name of the method is altered in the body by adding a version prefix and this is how we call the right version of a particular method.

So for example on route `/rpc/v0.2` and with the method in the body named `starknet_methodXyz` the body is transformed to contain `v0.2_starknet_methodXyz` because the RPC server is HTTP route-agnostic.

This brings us to ws, where you actually have to go though a protocol upgrade from http to ws, hence the empty body during handshake and the error that you encountered. Rightly so you added an exception to let the upgrade succeed but this still leaves us with the ws api essentially bypassing this current pathfinder api versioning ~~nightmare~~ scheme.

Reply:
Resolving as versioning is a matter for a separate pr